pkg/derivation/derivation.go

```diff
@@ -81,29 +86,182 @@ func (d *Derivation) Validate() error {
+		store = NewInputSubstKV(store)
+	}
+
+	name := d.Name()
+
+	var buf bytes.Buffer
+	{
+		err := d.writeDerivation(&buf, true, store)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var partialDrvHash string
+	{
+		h := sha256.New()
+
+		_, err := h.Write(buf.Bytes())
+		if err != nil {
+			return nil, err
+		}
+
+		partialDrvHash = hex.EncodeToString(h.Sum(nil))
+	}
+
+	outputs := make(map[string]string)
+
+	for _, o := range d.Outputs {
+		fixed := d.FixedOutput()
+		if fixed != nil {
+			s := fmt.Sprintf("source:sha256:%s:%s:%s", o.Hash, nixpath.StoreDir, name)
+
+			h := sha256.New()
+
+			_, err := h.Write([]byte(s))
+			if err != nil {
+				return nil, err
+			}
+
+			digest := h.Sum(nil)
+
+			outputs[o.Name] = filepath.Join(nixpath.StoreDir, nixbase32.EncodeToString(compressHash(digest, 20))+"-"+name)
+
+			continue
+		}
+
+		outputSuffix := name
+		if o.Name != "out" {
+			outputSuffix += "-" + o.Name
+		}
+
+		s := fmt.Sprintf("output:%s:sha256:%s:%s:%s", o.Name, partialDrvHash, nixpath.StoreDir, outputSuffix)
+
+		var digest []byte
+		{
+			h := sha256.New()
+
+			_, err := h.Write([]byte(s))
+			if err != nil {
+				return nil, err
+			}
+
+			digest = h.Sum(nil)
+		}
+
+		outputs[o.Name] = filepath.Join(nixpath.StoreDir, nixbase32.EncodeToString(compressHash(digest, 20))+"-"+outputSuffix)
+	}
+
+	return outputs, nil
+}
+
+// FixedOutput - Returns the fixed output any is found, otherwise returns nil.
+func (d *Derivation) FixedOutput() *Output {
+	for _, o := range d.Outputs {
+		if o.HashAlgorithm != "" {
+			return &o
+		}
+	}
+
+	return nil
+}
+
 // WriteDerivation writes the textual representation of the derivation to the passed writer.
 func (d *Derivation) WriteDerivation(writer io.Writer) error {
+	return d.writeDerivation(writer, false, nil)
+}
+
+func (d *Derivation) writeDerivation(writer io.Writer, maskOutputs bool, actualInputs KVStore) error {
 	outputs := make([][]byte, len(d.Outputs))
-	for i, o := range d.Outputs {
-		outputs[i] = encodeArray('(', ')', true, []byte(o.Name), []byte(o.Path), []byte(o.HashAlgorithm), []byte(o.Hash))
+	{
+		for i, o := range d.Outputs {
+			var path []byte
+			if maskOutputs {
+				path = []byte{}
+			} else {
+				path = []byte(o.Path)
+			}
+
+			outputs[i] = encodeArray('(', ')', true, []byte(o.Name), path, []byte(o.HashAlgorithm), []byte(o.Hash))
+		}
 	}
 
+	var err error
+
 	inputDerivations := make([][]byte, len(d.InputDerivations))
 	{
 		for i, in := range d.InputDerivations {
+			var path []byte
+			if actualInputs != nil {
+				path, err = actualInputs.Get(in.Path)
+				if err != nil {
+					return err
+				}
+
+				path = append([]byte{'"'}, path...) // TODO: Inefficient
+				path = append(path, '"')
+			} else {
+				path = quoteString(in.Path)
+			}
+
 			names := encodeArray('[', ']', true, stringsToBytes(in.Name)...)
-			inputDerivations[i] = encodeArray('(', ')', false, quoteString(in.Path), names)
+			inputDerivations[i] = encodeArray('(', ')', false, path, names)
```

Comment:
same as above, probably easier to just write two different blocks setting `inputDerivations[i]`, depending on whether `actualInputs` is nill or not.